[WIP] Improve artifact choice

barskern opened the pull request:
Improves the choice of artifact by searching through all the artifacts and ~finding the artifact which has the same package name in `package_id` as the name of the current running package~ then picking the one that is either `bin`, `example`, `test` or marked as a test build (`profile.test == true`). The user has to use `--bin`, `--example`, `--test` and `--lib` to disambiguate the binary to be examined.

This may seem to add a lot of code, but most of it is gracefull and informative error handling. ~It will also always fall back to taking the last artifact upon any error searching~.

~TODO: Currently the code searches using `CARGO_PKG_NAME` hence this env variable needs to be set. This has to be set manually using `CARGO_PKG_NAME=my-package`, which is inconvenient. I'm looking into a better way of getting the name.~

barskern committed:
Restructure crate into multiple modules

cbourjau commented:
Thanks yet again! I will take a closer look at this one tomorrow as it is getting quite late here in my neck of the world. But I agree that the location of the binary should be better handled indeed.

barskern commented:
I was trying to search for some crate that could read and parse the `Cargo.toml` of the current project where the subcommand is called, however I couldn't seem to find one. Hence I think we would need to try to find a `Cargo.toml` in all parent directories.

My current idea is to use [structopt_toml](https://docs.rs/structopt-toml/0.3.0/structopt_toml/index.html) to make the commandline arguments a bit more sophisticated. This way the program could read `Cargo.toml` on startup and fill the arguments with defaults, and those values could be overriden by the user if commandline arguments are specified. What do you think?

barskern commented:
There is also [`cargo_metadata`](https://github.com/oli-obk/cargo_metadata/blob/master/README.md) which could get the info from the `Cargo.toml`. In practical terms I think either solution will be quite similar.

cbourjau commented:
Sorry for taking a while to get back to you! reading up on `cargo_metadata` it seems like maybe there is additional information being emitted by cargo which is currently ignored. As it stands now `cargo-with` tries to parse the output from `cargo build --message-format=json` using `serde` into the struct `BuildOpt`. This parsing currently fails for a few lines and I have not yet looked into the reason for this. Maybe there is a way to fish out the package name from there?

barskern commented:
No worries man :smile:

Could you give an example of where it failes? Like do you mean that it failes to deserialize some of the fields or? From what I can see `cargo build --message-format=json` returns a bunch of top level json objects, which might be difficult to parse correctly.

Regarding getting the package name from that output, that is definetly possible. I think I actually found a nice way of doing it. It seems like the executable will be the only json struct where `kind` is set to `bin`. E.g:

```
  "target": {
    "kind": [
      "bin"
    ],
    "crate_types": [
      "bin"
    ],
    "name": "cargo-testify",
    "src_path": "/home/oruud/Programming/Github/cargo-testify/src/main.rs",
    "edition": "2015"
  },
```

cbourjau commented:
However, `cargo-with` should also work with the tests of a `lib`. Taking a closer look at the cargo output, I noticed that one should probably pass the quiet (`-q`) flag to reduce to avoid a mix up with human readable output. This is probably responsible for many of the json parsing errors I have seen. Furthermore, looking for artifacts which do **not** end in `.rlib` should be a strong indicator as well. I will add a few json outputs as testcases to demonstrate what I mean.

cbourjau commented:
Ok, I was playing around with this and realized that the file extension is probably very platform dependent! Should be reasonably easy to cover the three Tier 1 platforms, but I am not sure if it's worth the complexity? Maybe its better to wait on this and to see if there are cases in the wild where the current naive approach breaks?
For the record, the result of my playing around is in `origin/ICE`, but currently triggers the ICE described in https://github.com/rust-lang/rust/issues/55513

barskern commented:
Ah using the `-q` flag is smart. Previously I did `cargo build --message-format=json 2>/dev/null` to remove the human part of the output.

Yeah this not that nice to have something that's platform dependent. 

I've also been playing a bit around with it, and I think I found a solution that I would work. It's based on my previous suggestion of using `kind` from the target, but there are some more parameters we could use. There is also a `profile.test` which specifices if the output is a test binary. The way I've been playing with the JSON is through [`jq`](https://stedolan.github.io/jq/).

The following command processes the output into a ordered list where each "found" crate gets a label. It searches for `target.kind` which is `bin` and any executable marked as a `test` binary. When a binary has both, it will put a `test` label infront anyways.
```
cargo test --message-format=json -q \
  | rg '^[{]' \
  | jq 'select((.target.kind // [] | any(.=="bin")) or .profile.test) | if .profile.test then "test: \(.target.name)" else "bin: \(.target.name)" end'
```
This way we could make it both generic based on architecture (highly doubt that a binary is called anything but `bin` on other platforms) and work with libraries when running `cargo test`!

## Examples

Running the above command in `<pkg>` gives:

- `cargo-with`
```
test: cargo-with
```
- `bitflags`
```
test: bitflags
```
- `sonnerie` (first time)
```
bin: sonnerie
test: protocol
test: sonnerie
```
- `sonnerie` (second time, just wanted to highlight that they change order)
```
test: protocol
test: sonnerie
bin: sonnerie
```
- `serde`
```
test: serde_derive
test: serde_derive_internals
test: test_gen
test: test_annotations
test: serde_test
test: test_unstable
test: compiletest
test: test_borrow
test: test_ser
test: serde
test: test_identifier
test: test_remote
test: test_roundtrip
test: test_macros
test: test_value
test: test_de
```

Running the same command with `build` instead of test gives:

- `cargo-with`
```
bin: cargo-with
```
- `sonnerie`
```
bin: sonnerie
```

You get the point, it seems like a "better" solution than picking the last, but it will be quite simple to implement I think.

The only difficulty is that when running `cargo test` in a crate that has integration tests, is that there are multiple test binaries being built. This can be limited using `cargo test --test <integration-name>`, but we should probably get that choice from the user. I have two suggestions for dealing with that:

1. Show a menu and let the user choose
2. Throw an error stating that the user has to specify which test binary should be examined

Personally I favour the latter.

What are your thoughts? :smile:

cbourjau commented:
Oh boy, things are never as easy as they seem! Clearly, its quite common to have several binaries pop out at the end. In that case, I also prefer the second case prompting the user for more input. Looking at the `cargo` [source](https://github.com/rust-lang/cargo/blob/489f570d47dea88d8f9a1205173d3e779e1a78e2/src/cargo/core/manifest.rs#L239-L244), it seems like a third kind of `kind` is produced when running examples. I don't have a project with examples at hand at the moment, but hope to fine the time to look into this next week. Of course I'd be super happy to also merge your PR in this regard :D

barskern commented:
Ah yeah we should definitely think about examples aswell. I have some project with examples which I'll try it on. I think that the example problem could probably also be solved but `cargo run --example <example-name>`, so finding multiple examples should then just result in a error or prompt for specific example.

We should remember though that *normal* users who only have small crates will never see these prompts, so I don't think they add any unnecessary complexity, which is great. 👍

barskern commented:
Compiling examples produces very predictable output:

```
cargo test --examples --message-format=json -q --release | rg '^[{]' | jq -r '"\(.target.kind[0]), \(.target.name)"'
```

```
...
example, window
example, pong_tutorial_01
example, sphere
example, pong_tutorial_02
example, asset_loading
example, net_client
example, animation
example, pong
example, sphere_multisample
example, separate_sphere
example, fly_camera
example, pong_tutorial_03
example, prefab
example, sprites_ordered
example, hello_world
example, custom_game_data
example, net_server
example, locale
example, material
example, spotlights
example, debug_lines
example, renderable
example, custom_ui
example, appendix_a
example, ui
example, arc_ball_camera
example, gltf
```

barskern commented:
So basically we are interested in `bin`, `example` and `test` target kinds, which means we could filter the output based on that. If there is more than one possible candidate, we could return an error message saying that the user needs to be more specific. Doesn't seem to hard to make :smile:

barskern committed:
Improve selection of artifact

- Uses `.target.kind` and `.profile.test` to find the candidates.
- There can only be one candidate.
- If more than one candidate, ask the user to use `--test`, `--example`,
  `--bin` or `--lib` to specify binary.

barskern commented:
This PR now encompasses a lot more than I initially imagined. I have now implemented a way of chosing artifacts based on the output from the cargo subcommand using `target.kind` and `profile.test`. Based on the projects I tested it on, it should work pretty well on libraries, examples, unit-tests and binaries! Normally the user has to specify either `--lib`, `--bin`, `--example` or `--test` to make it work, but that will only affect cases where there are multiple possible binaries the user might want to examine. Some more testing is probably beneficial before it should be considered mergable. 

I also ended up removing `{args}` because if the user want to specify args to pass to the executable it's easier if its done directly in the `<with-cmd>` parameter. This could of course be reverted if you want to have `{args}` back. :smile:

cbourjau commented:
Oh wow, thanks again! Sorry, I was on holidays in the meantime. There is a lot going on here but I very much like how it is shaping up! Lets start with when `cargo-with` should actually work. I suppose the obvious ones are `cargo test`, `cargo run`, and `cargo run --example my_example`. Of these, I think only the first one may produces ambiguities, does it not? The case of `cargo test --example` seems to be a little esoteric, as even the [cargo maintainers](https://github.com/rust-lang/cargo/issues/3559) suggested. Thus, I think asking the user for clarification in that case would be fine, I think.

I suppose another use case might be that the user compiles a `lib` and wants to do something with the produced artifact (ie, not run but maybe do some analysis on it). In this case, `cargo run` would not suffice and `cargo build` would be the logical addition. Do you think we should support `cargo build`?

I don't quite understand what you mean with 

> Normally the user has to specify either --lib, --bin, --example or --test to make it work

Maybe I don't understand correctly, but I don't see how this resolves ambiguities? Don't we already know whether its a `lib`, `bin`, or `test` we are looking for?

Lastly, there is the question about `{args}`. The reason I added this was due to the following use case. The user wants to debug a binary which is executed with arguments. So the standard cargo command is:

```
cargo run -- test_args
```
Ie, the user already provided the arguments and it would be inconvenient to copy past them around to the string supplied to `cargo-with`. Always appending them to the end might also not work, since other arguments might have to follow the execution arguments in the case of the debugger. So `gdb` for example expects the arguments to the executable as:
```
gdb {bin} --args {args}
```
Do you agree that there is value in having `{args}`?

In any case, thanks so much for putting the extra time into this to improve this tool!

cbourjau commented:
Oh, and for future reference to myself and others: a project with plenty of  `examples` is ['clap'](https://github.com/clap-rs/clap/tree/master/examples)

barskern commented:
No worries man, holidays are important :smile:

Yeah the main goal was `cargo run` and all forms of `cargo test`. Its `cargo test` that actually brings the need for explicitness from the user. Running `cargo test --help`, you see that you can specify `--bin`, `--lib`, `--test` and `--example`, which makes it hard to determine which package the user wants to test without specifying it explicitly. Hence what I meant with "Normally the use has to specify either [...]", is that when running `cargo test` there could be ambiguity. If you have a `lib.rs` and a `main.rs` you need to specify either `--lib <package-name>` or `--bin <package-name>` to specify whether you want to analyze the test build for the `main.rs` file and its modules or the `lib.rs` file and its modules. As it seems to be quite normal to use both `lib.rs` and `main.rs` in small executables, every user that wants to use `cargo --test` has to be more specific than just `cargo --test`.

In the output from `cargo build --message-format=json` there is not a clean way of disambiguating the current library from another dependency. The only parameter I think we could use to distinguish the current library from the dependency is the source path. We could compare the current directory to the source path of the different libraries, and then pick the library which has it's root in the current directory. However this seems like a hacky solution that will probably break easily. We could solve this with our own CLI argument to specify the name of the library the user would like to examine.

I did not think about that usecase of `{args}`, but that is actually a quite ergonomic way of doing it. I think it should be re-added. I will commit the changes when I have time.

barskern commented:
`{args}` is now added back in :smile: Perhaps the solution is a bit overengineered using `impl Trait` etc. My reasoning was to make it easy to expand on later on, and to prevent as many `Vec` creations as possible as a personal exercise. What do you think?

~PS! Perhaps I should squash first? Some of the commits are small and there are some redundancy ("improve error message" etc..).~ Squashed and ready for merge :+1:

cbourjau commented:
This looks great! Thanks again for digging into this so deep and I know all to well the feeling of wanting to overengineered things with all the cool rust features ;) I just merged this and also made you a collaborator for the future. I hope you don't mind. I'll push an update to crates.io soon as well. I thought about improving the readme a bit first, though.

cbourjau commented:
Oh, and concerning the example with `gdb` needing `{args}`: I'm sorry to admit that I got that wrong! Actually, the correct way of calling gdb with args for the inferior is (as actually given in the readme...):

    $ gdb --args my_bin arg1 arg2

Sorry about that confusion! I still think that having such a flexibility for other executables might be the right thing, though.

barskern commented:
I could summarize the new way of choosing binary :smile:

cbourjau commented:
@barskern That would be great! Depending on how long you want to make that description, we could also link to it from the readme, to keep the readme concise and user-focused.
